src/models/FirestoreModel.ts

```diff
@@ -0,0 +1,50 @@
+import { firestore } from 'firebase';
+import Model from './Model';
+
+export default class FirestoreModel<T> implements Model<T> {
+  constructor(private readonly ref: firestore.DocumentReference<T>, public readonly data: T) {}
+
+  public get id(): string {
+    return this.ref.id;
+  }
+
+  get(): Promise<FirestoreModel<T>> {
+    return new Promise((resolve, reject) => {
+      this.ref.get()
+        .then((doc: firestore.DocumentSnapshot<T>) => {
+          if (!doc.exists) {
+            reject(new Error(`${this.ref.path} not found`));
+            return;
+          }
+          resolve(new FirestoreModel<T>(doc.ref, doc.data() as T));
```

Comment:
`doc.data() as T`
Would this interpret the data as our class models? e.g. User

Reply:
Yes that's right